sourcegraph/sourcegraph
# grpc: searcher: change file_path and content type to be bytes instead of string

On sourcegraph.com, we noticed grpc errors on Sunday that were complaining that searcher gRPC service was returning file paths that aren't utf-8 encoded. See [this log on sourcegraph.com](https://console.cloud.google.com/logs/query;cursorTimestamp=2023-10-02T02:59:22.400900821Z;duration=P7D;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22sourcegraph-dev%22%0Aresource.labels.location%3D%22us-central1-f%22%0Aresource.labels.cluster_name%3D%22cloud%22%0Aresource.labels.namespace_name%3D%22prod%22%0AjsonPayload.InstrumentationScope%3D~%22gRPC.internal.error.reporter%22%0Atimestamp%3D%222023-10-02T02:59:22.372818886Z%22%0AinsertId%3D%222buyn24bv6tl4hvl%22?project=sourcegraph-dev):

```

{
  "insertId": "2buyn24bv6tl4hvl",
  "jsonPayload": {
    "Attributes": {
      "grpcService": "searcher.v1.SearcherService",
      "nonUTF8StringFields": [
        "file_match.path"
      ],
      "grpcMethod": "Search",
      "initialRequestJSON": "{\"repo\":\"github.com/pde/tosback2-data\",\"repo_id\":57116556,\"commit_oid\":\"15c5d28f3bd989e235ccf5b0135c73950a02dd30\",\"pattern_info\":{\"is_regexp\":true,\"include_patterns\":[\"wp-content/\"],\"limit\":99999999,\"pattern_matches_content\":true,\"pattern_matches_path\":true},\"fetch_timeout\":{\"seconds\":15,\"nanos\":716557393},\"feat_hybrid\":true}",
      "messageJSON": "{\"Message\":{\"FileMatch\":{\"path\":\"crawls/globalresearch.ca/Privacy-Policy/raw/www.globalresearch.ca/wp-content/uploads/2013/04/\\ufffd%89lections-Venezuela-Maduro-51x46.jpg\"}}}",
      "grpcCode": "Internal"
    },
    "Function": "github.com/sourcegraph/sourcegraph/internal/grpc/internalerrs.doLog",
    "Caller": "internalerrs/logging.go:240",
    "InstrumentationScope": "searcher.gRPC.internal.error.reporter.streamingMethod.postMessageSend",
    "message": "grpc: error while marshaling: string field contains invalid UTF-8",
    "timestampNanos": 1696215562372458000,
}
```

In protobuf, all strings must be utf-8 encoded according to the spec: https://protobuf.dev/programming-guides/proto3/#scalar. If you can't be sure that your strings are utf-8 strings, you must relax the type to use bytes instead.

This has been occuring for a while on sourcegraph.com, we were never validating / santizing file paths from respositories. As such, some behavior has probably been silently breaking for a while due to this:  https://github.com/sourcegraph/sourcegraph/blob/d73bf9853c4a750353fec5941bf067563d7d67b2/internal/search/searcher/search.go#L332-L343

This PR papers over the issue by relaxing both the file path and content types to use `bytes` instead of `string`. I've added casts in our conversion functions to convert them back to strings. This is not a true fix for the issue (we should be sanitizing this input at the edges of our application, but it works around this problem for now). 

## Test plan

Manually tested:

1. Clone github.com/pde/tosback2-data to your local machine
2. Run the following search: https://sourcegraph.test:3443/search?q=context:global+repo:%5Egithub%5C.com/pde/tosback2-data%24%4015c5d28+file:%5Ecrawls/globalresearch%5C.ca/Privacy-Policy/raw/www%5C.globalresearch%5C.ca/wp-content/uploads/2013/04+f:maduro&patternType=regexp&sm=1&groupBy=path

3. See that it returns valid results instead of a search error: ![Screenshot 2023-10-02 at 3 03 26 PM](https://github.com/sourcegraph/sourcegraph/assets/9022011/464c0e06-c6ac-4394-a864-2da1df657d92)
- `crawls/globalresearch.ca/Privacy-Policy/raw/www.globalresearch.ca/wp-content/uploads/2013/04/�%89lections-Venezuela-Maduro-51x46.jpg` was the non-utf8 file name, notice that it appears in the search results
	- Running the same search with the version of searcher prior to this fix leads to a missing file:
			![Screenshot 2023-10-02 at 3 06 48 PM](https://github.com/sourcegraph/sourcegraph/assets/9022011/6207d8dd-d092-4129-a77b-b6c02990a41b)
